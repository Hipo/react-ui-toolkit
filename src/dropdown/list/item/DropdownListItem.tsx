import "./_dropdown-list-item.scss";

import React, {Fragment} from "react";
import classNames from "classnames";

export interface DropdownOption<Id = string, Context = any> {
  id: Id;
  title: string;
  customClassName?: string;
  CustomContent?: JSX.Element;
  icon?: React.ReactNode;
  subtitle?: string;
  context?: Context;
  isDisabled?: boolean;
}

export type DropdownOptionSelectHandler<Id = string, Context = any> = (
  option: DropdownOption<Id, Context> | null,
  event?: React.SyntheticEvent<HTMLLIElement>
) => void;

export type DropdownSelectedOption<Id = string, Context = any> =
  | DropdownOption<Id, Context>
  | null
  | undefined;

export interface DropdownListItemProps<OptionIdShape = string> {
  testid?: string;
  option: DropdownOption<OptionIdShape>;
  selectedOption: DropdownSelectedOption<OptionIdShape>;
  focusedOption?: DropdownSelectedOption<OptionIdShape>;
  onSelect: DropdownOptionSelectHandler<OptionIdShape>;
  onFocus: DropdownOptionSelectHandler<OptionIdShape>;
  onKeyDown?: DropdownOptionSelectHandler<OptionIdShape>;
  onMouseDown?: React.ReactEventHandler<HTMLLIElement>;
  onMouseUp?: React.ReactEventHandler<HTMLLIElement>;
  canSelectAlreadySelected?: boolean;
  focusedItemRef?: React.RefObject<HTMLLIElement>;
}

function DropdownListItem<OptionIdShape extends string>({
  testid,
  option,
  selectedOption,
  onSelect,
  focusedOption,
  onFocus,
  onKeyDown,
  onMouseDown,
  onMouseUp,
  canSelectAlreadySelected = false,
  focusedItemRef
}: DropdownListItemProps<OptionIdShape>) {
  const {id: optionId, customClassName, icon, title, subtitle, CustomContent} = option;
  const isSelected = Boolean(selectedOption && optionId === selectedOption.id);
  const canItemBeClicked =
    !option.isDisabled && (!isSelected || canSelectAlreadySelected);
  const containerClassName = classNames("dropdown-list-item", customClassName, {
    "dropdown-list-item--is-selected": isSelected,
    "dropdown-list-item--is-focused": Boolean(
      focusedOption && optionId === focusedOption.id
    ),
    "dropdown-list-item--can-be-selected": canItemBeClicked,
    "dropdown-list-item--is-disabled": option.isDisabled
  });

  return (
    <li
      ref={focusedItemRef}
      data-testid={testid}
      id={optionId}
      role={"option"}
      className={containerClassName}
      onClick={handleClick}
      onFocus={handleFocus}
      onMouseEnter={handleFocus}
      onMouseDown={onMouseDown}
      onMouseUp={onMouseUp}
      onKeyDown={handleKeyDown}
      aria-selected={isSelected}>
      {CustomContent || (
        <Fragment>
          <div className={"dropdown-list-item__icon"}>{icon}</div>

          <div className={"dropdown-list-item__content"}>
            <p data-testid={`${testid}.title`} className={"dropdown-list-item__title"}>
              {title}
            </p>

            {subtitle && (
              <p
                data-testid={`${testid}.subtitle`}
                className={"dropdown-list-item__subtitle"}>
                {subtitle}
              </p>
            )}
          </div>
        </Fragment>
      )}
    </li>
  );

  function handleClick(event: React.SyntheticEvent<HTMLLIElement>) {
    event.preventDefault();
    event.stopPropagation();

    if (canItemBeClicked) {
      onSelect(option, event);
    }
  }

  function handleFocus(event: React.SyntheticEvent<HTMLLIElement>) {
    onFocus(option, event);
  }

  function handleKeyDown(event: React.KeyboardEvent<HTMLLIElement>) {
    if (onKeyDown) {
      onKeyDown(option, event);
    }
  }
}

export default DropdownListItem;
