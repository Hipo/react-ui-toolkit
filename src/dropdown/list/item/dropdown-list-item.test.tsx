import React from "react";
import {render, screen, fireEvent} from "@testing-library/react";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";

import {testA11y} from "../../../core/utils/test/testUtils";
import DropdownListItem, {DropdownListItemProps} from "./DropdownListItem";

describe("<DropdownListItem />", () => {
  const defaultDropdownListItemProps: DropdownListItemProps = {
    testid: "dropdown-list-item",
    selectedOption: {id: "1", title: "test"},
    option: {id: "1", title: "test"},
    onFocus: jest.fn(),
    onSelect: jest.fn(),
    onKeyDown: jest.fn()
  };

  it("should render correctly", () => {
    render(<DropdownListItem {...defaultDropdownListItemProps} />);
  });

  it("should pass a11y test", async () => {
    const {container} = render(<DropdownListItem {...defaultDropdownListItemProps} />);

    await testA11y(container, {rules: {"aria-required-parent": {enabled: false}}});
  });

  it("should have proper modifier class name", () => {
    const {rerender} = render(
      <DropdownListItem
        {...defaultDropdownListItemProps}
        canSelectAlreadySelected={true}
        focusedOption={{...defaultDropdownListItemProps.option}}
      />
    );

    expect(screen.getByRole("option")).toHaveClass("dropdown-list-item--is-selected");
    expect(screen.getByRole("option")).toHaveClass("dropdown-list-item--is-focused");
    expect(screen.getByRole("option")).toHaveClass("dropdown-list-item--can-be-selected");

    rerender(
      <DropdownListItem
        {...defaultDropdownListItemProps}
        option={{...defaultDropdownListItemProps.option, isDisabled: true}}
      />
    );

    expect(screen.getByRole("option")).toHaveClass("dropdown-list-item--is-disabled");
  });

  it("should render custom content if it exists", () => {
    const {container} = render(
      <DropdownListItem
        {...defaultDropdownListItemProps}
        option={{
          ...defaultDropdownListItemProps.option,
          CustomContent: <p>{"custom content"}</p>
        }}
      />
    );

    expect(container).toContainElement(screen.getByText("custom content"));
  });

  describe("should handle event handlers correctly", () => {
    it("should not run click event handler when option is disabled and option is selected or canSelectAlreadySelected is provided as false", () => {
      // isDisabled: true, isSelected: true , canSelectAlreadySelected: default = false
      const {rerender} = render(
        <DropdownListItem
          {...defaultDropdownListItemProps}
          option={{...defaultDropdownListItemProps.option, isDisabled: true}}
        />
      );

      userEvent.click(screen.getByRole("option"));
      expect(defaultDropdownListItemProps.onSelect).not.toHaveBeenCalled();

      // isDisabled: true, isSelected: false , canSelectAlreadySelected: default = false
      rerender(
        <DropdownListItem
          {...defaultDropdownListItemProps}
          selectedOption={{id: "2", title: "test 2"}}
          option={{...defaultDropdownListItemProps.option, isDisabled: true}}
        />
      );

      userEvent.click(screen.getByRole("option"));
      expect(defaultDropdownListItemProps.onSelect).not.toHaveBeenCalled();

      // isDisabled: true, isSelected: true , canSelectAlreadySelected: true
      rerender(
        <DropdownListItem
          {...defaultDropdownListItemProps}
          canSelectAlreadySelected={true}
          option={{...defaultDropdownListItemProps.option, isDisabled: true}}
        />
      );

      userEvent.click(screen.getByRole("option"));
      expect(defaultDropdownListItemProps.onSelect).not.toHaveBeenCalled();
    });

    it("should run click event handler when option is not disabled and option is not selected or canSelectAlreadySelected is provided as true", () => {
      // isDisabled: false, isSelected: true , canSelectAlreadySelected: true
      const {rerender} = render(
        <DropdownListItem
          {...defaultDropdownListItemProps}
          canSelectAlreadySelected={true}
        />
      );

      userEvent.click(screen.getByRole("option"));
      expect(defaultDropdownListItemProps.onSelect).toHaveBeenCalledTimes(1);

      jest.clearAllMocks();

      // isDisabled: false, isSelected: false , canSelectAlreadySelected: default = false
      rerender(
        <DropdownListItem
          {...defaultDropdownListItemProps}
          selectedOption={{id: "2", title: "test 2"}}
        />
      );

      userEvent.click(screen.getByRole("option"));
      expect(defaultDropdownListItemProps.onSelect).toHaveBeenCalledTimes(1);

      jest.clearAllMocks();

      // isDisabled: false, isSelected: false , canSelectAlreadySelected: true
      rerender(
        <DropdownListItem
          {...defaultDropdownListItemProps}
          selectedOption={{id: "2", title: "test 2"}}
          canSelectAlreadySelected={true}
        />
      );

      userEvent.click(screen.getByRole("option"));
      expect(defaultDropdownListItemProps.onSelect).toHaveBeenCalledTimes(1);
    });

    it("should run focus event handler correctly", () => {
      const {container} = render(<DropdownListItem {...defaultDropdownListItemProps} />);

      fireEvent.focus(container);

      expect(defaultDropdownListItemProps.onFocus).toHaveBeenCalledTimes(1);
    });

    it("should run key down event handler correctly", () => {
      const {rerender} = render(<DropdownListItem {...defaultDropdownListItemProps} />);

      fireEvent.keyDown(screen.getByRole("option"), {
        keyCode: 16
      });

      expect(defaultDropdownListItemProps.onKeyDown).toHaveBeenCalledTimes(1);

      jest.clearAllMocks();

      rerender(
        <DropdownListItem {...defaultDropdownListItemProps} onKeyDown={undefined} />
      );

      expect(defaultDropdownListItemProps.onKeyDown).not.toHaveBeenCalled();
    });
  });
});
