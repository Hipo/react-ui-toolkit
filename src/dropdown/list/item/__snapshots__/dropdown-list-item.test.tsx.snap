// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropdownListItem /> should match snapshot 1`] = `
<li
  aria-selected={true}
  className="dropdown-list-item dropdown-list-item--is-selected"
  data-testid="dropdown-list-item"
  id="1"
  onClick={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onMouseEnter={[Function]}
  role="option"
>
  <div
    className="dropdown-list-item__icon"
  />
  <div
    className="dropdown-list-item__content"
  >
    <p
      className="dropdown-list-item__title"
      data-testid="dropdown-list-item.title"
    >
      test
    </p>
  </div>
</li>
`;
